**Autores**: 
1. Caio Fernandes - 170138798
2. Lucas Dutras - 170050939

Este projeto tem como objetivo utilizar técnicas de data science juntamente com o conteúdo de **Percorrer Grafos** para indicar qual o caminho mais rápido para chegar em um país a partir de um país inicial.

Utilizaremos para isso dois datasets disponibilizados no Kaggle com informações sobre [Países e Fronteiras](https://www.kaggle.com/resheto/country-borders) e [Países e Continentes](https://www.kaggle.com/tomvebrcz/countriesandcontinents).

In [133]:
import pandas as pd

In [134]:
df_border = pd.read_csv('./data/border_info.csv')
df_continents = pd.read_csv('./data/countries.csv')
df_border = df_border.sort_values(by=['country_name'])
df_continents = df_continents.sort_values(by=['Country'])

## Avaliação do Dado

In [135]:
df_border.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 728 entries, 9 to 727
Data columns (total 3 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Unnamed: 0           728 non-null    int64 
 1   country_name         728 non-null    object
 2   country_border_name  642 non-null    object
dtypes: int64(1), object(2)
memory usage: 22.8+ KB


## Limpando o dado 

Removendo países sem fronteira

In [136]:
is_NaN = df_border.isnull()
row_has_NaN = is_NaN.any(axis=1)
rows_with_NaN = df_border[row_has_NaN]
rows_with_NaN

Unnamed: 0                          country_name country_border_name
41           41                         Aland Islands                 NaN
30           30                        American Samoa                 NaN
11           11                              Anguilla                 NaN
24           24                            Antarctica                 NaN
10           10                   Antigua and Barbuda                 NaN
..          ...                                   ...                 ...
684         684  United States Minor Outlying Islands                 NaN
704         704                               Vanuatu                 NaN
699         699              Virgin Islands (British)                 NaN
700         700                 Virgin Islands (U.S.)                 NaN
705         705                     Wallis and Futuna                 NaN

[86 rows x 3 columns]

Temos 86 países sem fronteiras no dado

Removendo-os:

In [137]:
df_border = df_border.drop(rows_with_NaN.index)

Notamos que, neste dado em questão, temos 727 valores.
Tal dado também apresenta duas colunas, uma `country_name` representando o país e outra chamada `coutry_border_name` que representa um fronteira daquele país.

In [138]:
df_border.head()

Unnamed: 0 country_name country_border_name
9           9  Afghanistan          Uzbekistan
4           4  Afghanistan               China
5           5  Afghanistan                Iran
6           6  Afghanistan            Pakistan
7           7  Afghanistan          Tajikistan

Logo percebemos um problema, a maneira como o dado está organizado, nos trás informações duplicadas sobre países.
Para corrigir este problema, faremos, para cada país, uma coluna com uma lista de todas as fronteiras.
Com isso, teremos na estrutura do `DataFrame`, países únicos na coluna `country_name` com suas fronteiras listada na coluna `border_list`.

In [139]:
k = df_border.groupby('country_name')['country_border_name'].apply(lambda x: list(x))
df_border = df_border.drop(columns=['country_border_name'])
df_border = df_border.drop_duplicates(subset=['country_name'])
aux = [item for item in k]
df_border['border_list'] = aux

In [141]:
df_border = df_border.reset_index()
df_border = df_border.drop(columns=['Unnamed: 0', 'index'])
df_continents['Region'] = df_continents['Region'].replace('LATIN', 'AMERICA')
df_continents['Region'] = df_continents['Region'].replace('NORTHERN_AMERICA', 'AMERICA')
df_continents['Region'] = df_continents['Region'].replace('NEAR_EAST', 'ASIA')

In [142]:
border_l = df_border['country_name'].tolist()
cont_l = df_continents['Country'].tolist()
diff = []
for item in cont_l:
    if item in border_l:
        continent = df_continents.loc[df_continents['Country'] == item]
        continent = continent['Region']
        for elem in continent.iteritems():
            df_border.loc[df_border['country_name'] == item, 'continent'] = elem[1]


In [143]:
df_border.head(20)

country_name  \
0                              Afghanistan   
1                                  Albania   
2                                  Algeria   
3                                  Andorra   
4                                   Angola   
5                                Argentina   
6                                  Armenia   
7                                  Austria   
8                               Azerbaijan   
9                               Bangladesh   
10                                 Belarus   
11                                 Belgium   
12                                  Belize   
13                                   Benin   
14                                  Bhutan   
15                                 Bolivia   
16                  Bosnia and Herzegovina   
17                                Botswana   
18                                  Brazil   
19                                  Brunei   
20                                Bulgaria   
21                            Burkina Faso   
22                                 Burundi   
23                                Cambodia   
24                                Cameroon   
25                                  Canada   
26                Central African Republic   
27                                    Chad   
28                                   Chile   
29                                   China   
30                                Colombia   
31                        Congo (Kinshasa)   
32  Congo (the Democratic Republic of the)   
33                              Costa Rica   
34                           Cote d'Ivoire   
35                                 Croatia   
36                                 Czechia   
37                                 Denmark   
38                                Djibouti   
39                      Dominican Republic   
40                                 Ecuador   
41                                   Egypt   
42                             El Salvador   
43                       Equatorial Guinea   
44                                 Eritrea   
45                                 Estonia   
46                                Eswatini   
47                                Ethiopia   
48                                 Finland   
49                                  France   
50                           French Guiana   
51                                   Gabon   

                                          border_list continent  
0   [Uzbekistan, China, Iran, Pakistan, Tajikistan...      ASIA  
1       [Serbia, North Macedonia, Montenegro, Greece]    EUROPE  
2   [Mali, Tunisia, Niger, Libya, Morocco, Western...    AFRICA  
3                                     [France, Spain]    EUROPE  
4   [Congo (Kinshasa), Congo (the Democratic Repub...    AFRICA  
5         [Uruguay, Brazil, Bolivia, Chile, Paraguay]   AMERICA  
6                 [Georgia, Iran, Turkey, Azerbaijan]    EUROPE  
7   [Slovenia, Czechia, Germany, Hungary, Liechten...    EUROPE  
8            [Georgia, Turkey, Armenia, Iran, Russia]    EUROPE  
9                                    [Myanmar, India]      ASIA  
10       [Poland, Latvia, Lithuania, Russia, Ukraine]    EUROPE  
11         [Germany, Luxembourg, Netherlands, France]    EUROPE  
12                                [Guatemala, Mexico]   AMERICA  
13               [Burkina Faso, Niger, Nigeria, Togo]    AFRICA  
14                                     [India, China]      ASIA  
15         [Brazil, Peru, Paraguay, Chile, Argentina]   AMERICA  
16                      [Croatia, Serbia, Montenegro]       NaN  
17          [Zimbabwe, Zambia, South Africa, Namibia]    AFRICA  
18  [Venezuela, Suriname, Uruguay, Paraguay, Guyan...   AMERICA  
19                                         [Malaysia]      ASIA  
20  [Turkey, Serbia, Greece, Romania, North Macedo...    EUROPE  
21   [Cote d'Ivoire, Togo, Niger, Mali, Ghana, Benin]    AFRICA  
22  [Rwanda, Congo (the Democratic Republic of the...    AFRICA  
23     

In [144]:
df_continents['Region'].value_counts()

AFRICA     57
EUROPE     50
AMERICA    50
ASIA       49
OCEANIA    21
Name: Region, dtype: int64